Python/import.c

```diff
@@ -303,10 +303,120 @@ _PyImport_Fini(void)
 PyObject *
 PyImport_GetModuleDict(void)
 {
-    PyInterpreterState *interp = PyThreadState_GET()->interp;
-    if (interp->modules == NULL)
-        Py_FatalError("PyImport_GetModuleDict: no module dictionary!");
-    return interp->modules;
+    PyObject *sysdict = PyThreadState_GET()->interp->sysdict;
+    if (sysdict == NULL)
+        Py_FatalError("PyImport_GetModuleDict: no sys module!");
+
+    _Py_IDENTIFIER(modules);
+    PyObject *modules = _PyDict_GetItemId(sysdict, &PyId_modules);
+    if (modules == NULL)
+        Py_FatalError("lost sys.modules");
+    return modules;
+}
+
+/* In some corner cases it is important to be sure that the import
+   machinery has been initialized (or not cleaned up yet).  For
+   example, see issue #4236 and PyModule_Create2(). */
+
+int
+_PyImport_IsInitialized(PyInterpreterState *interp)
+{
+    if (interp->sysdict == NULL)
+        return 0;
+    _Py_IDENTIFIER(modules);
+    PyObject *modules = _PyDict_GetItemId(interp->sysdict, &PyId_modules);
+    if (modules == NULL)
+        return 0;
+    return 1;
+}
+
+PyObject *
+_PyImport_GetModule(PyObject *name)
+{
+    PyObject *modules = PyImport_GetModuleDict();
+    if (PyDict_Check(modules))
+        return PyDict_GetItem(modules, name);
+
+    PyObject *mod = PyObject_GetItem(modules, name);
+    // For backward-comaptibility we copy the behavior of PyDict_GetItem().
+    if (PyErr_Occurred())
+        PyErr_Clear();
+    Py_XDECREF(mod);
+    return mod;
+}
+
+PyObject *
+_PyImport_GetModuleWithError(PyObject *name)
```

Comment:
Is it worth it to have two implementations, _PyImport_GetModuleWithError() and _PyImport_GetModule()? Is it for best performances? Or can we implement _PyImport_GetModule() using _PyImport_GetModuleWithError()?

Reply:
I did it to mirror PyDict_GetItem() and PyDict_GetItemWithError().  I expect that there is a significant performance difference, but I'm not convinced that it would make a difference for real Python programs.  If you feel strongly about the duplication then I'll sort it out, but I'd rather not. :)

Reply:
`PyDict_GetItem()` is slower then `PyDict_GetItemWithError()` and its behavior looks weird to me (it can silence KeyboardInterrupt or programming errors). I consider it as legacy API and bad example for new API.